dask/base.py

```diff
@@ -928,6 +929,9 @@ def normalize_object(o):
     if callable(o):
         return normalize_function(o)
 
+    if dataclasses.is_dataclass(o):
+        return normalize_token(dataclasses.astuple(o))
```

Comment:
This seems to only take into account the values of dataclass, not the field names. For example, 

```python
import dataclasses
from dask.base import tokenize

# Create very similar dataclasses with different field names (i.e. "a" and "b") 
dataclass_1 = dataclasses.make_dataclass("dataclass_1", [("a", int)])
dataclass_2 = dataclasses.make_dataclass("dataclass_2", [("b", int)])
foo = dataclass_1(1)
bar = dataclass_2(1)
print(f"{tokenize(foo) = }")
print(f"{tokenize(bar) = }")
```

returns random tokens on `main` (this isn't ideal, but won't lead to incorrect computations), while with the changes in this PR we get identical tokens:

```
tokenize(foo) = 'd28a26e72c5cf6667e66de5a5e92bd62'
tokenize(bar) = 'd28a26e72c5cf6667e66de5a5e92bd62'
```

I'm not super familiar with dataclasses, but I would think we would want to incorporate field names too when generating deterministic hashes

Reply:
Great point. I can incorporate the field names. I'm debating whether the type of the dataclass should be included as well; if you have two dataclasses with the same field names and values, but different container types, should they be hashed differently? I also notice
1. [`astuple`](https://docs.python.org/3/library/dataclasses.html#dataclasses.astuple) makes a deep copy; we don't need that for tokenization
2. If the dataclass instance defines a `__hash__` method, we should probably respect that?

Reply:
Perhaps [`dataclasses.asdict`](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict) is more what we're looking for

Reply:
> I'm debating whether the type of the dataclass should be included as well

Just to clarify, this would be `dataclass_1` vs. `dataclass_2` in the above example? If so, I would recommend including that in the token as well. This is similar to [what we do with other similar-ish objects like `tuple` vs. `list`](https://github.com/dask/dask/blob/85bc7b5bb888c8a96eee3db07b260ca11d7cc130/dask/base.py#L909) 

> If the dataclass instance defines a __hash__ method, we should probably respect that?

Good question. Is that something users of dataclasses commonly do? IIRC we don't use `__hash__` in tokenizing other objects

Reply:
Yes, though I'd want to use the shallow-copy snippet they give there. I'm still a little uncertain about `__hash__` and the parent type though. I'm going to look at this a little more.

Reply:
> this would be `dataclass_1` vs. `dataclass_2` in the above example

Well those actually have different field names. I was more thinking
```python
ADataClass = dataclasses.make_dataclass("ADataClass", [("a", int)])
BDataClass = dataclasses.make_dataclass("BDataClass", [("a", Union[int, float])])

a1 = ADataClass(1)
b1 = BDataClass(1)
```
In that case, even though they have the same value and same field names, user code could theoretically change behavior depending on `isinstance(x, ADataClass)`.

> Is that something users of dataclasses commonly do? IIRC we don't use `__hash__` in tokenizing other objects

Dataclasses have a bit of logic around when and how to add a `__hash__` method (normally it's generated automatically, though you can define it if you wish). See the `unsafe_hash` section of the [docs](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass). Since you can specify [whether certain fields contribute to the hash](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) and/or to equality comparison, and all that gets taken into account in the autogenerated `__hash__` method, it feels to me like there are enough semantics around it that we should respect it.

As a user, if I explicitly set a dataclass field as `compare=False`, I think I'd expect dask to also ignore it when tokenizing? And if I wrote my own hash function, I think I'd also expect it to be taken into account.

In the end, tokenization is really meant as a form of equality comparison (are these inputs equivalent? are these results equivalent?), so I'd expect it to be consistent with any semantics around equality—which for dataclasses, hashing seems intentionally derived from. It's not quite equivalent, but might be connected enough to be a substitute for writing out the logic ourselves?